src/run.js

```diff
@@ -75,283 +75,293 @@ const minimalcss = async options => {
-          !link.href.toLowerCase().startsWith('blob:') &&
-          link.media !== 'print'
-        ) {
-          // if (!stylesheetAstObjects[link.href]) {
-          //   throw new Error(`${link.href} not in stylesheetAstObjects!`)
-          // }
-          // if (!Object.keys(stylesheetAstObjects[link.href]).length) {
-          //   // If the 'stylesheetAstObjects[link.href]' thing is an
-          //   // empty object, simply skip this link.
-          //   return
-          // }
-          hrefs.push(link.href)
-        }
+      // Second, goto the page and evaluate it on the 'networkidle2' event.
+      // This gives the page a chance to load any <script defer src="...">
+      // and even some JS that does XHR requests right after load.
+      response = await page.goto(pageUrl, {
+        waitUntil: ['domcontentloaded', 'networkidle2']
       })
-      return {
-        html: document.documentElement.outerHTML,
-        hrefs
+      if (!response.ok) {
+        return reject(
+          new Error(`${response.status} on ${pageUrl} (second time)`)
+        )
       }
-    })
+      const evalNetworkIdle = await page.evaluate(() => {
+        // The reason for NOT using a Set here is that that might not be
+        // supported in ES5.
+        const hrefs = []
+        // Loop over all the 'link' elements in the document and
+        // for each, collect the URL of all the ones we're going to assess.
+        Array.from(document.querySelectorAll('link')).forEach(link => {
+          if (
+            link.href &&
+            (link.rel === 'stylesheet' ||
+              link.href.toLowerCase().endsWith('.css')) &&
+            !link.href.toLowerCase().startsWith('blob:') &&
+            link.media !== 'print'
+          ) {
+            // if (!stylesheetAstObjects[link.href]) {
+            //   throw new Error(`${link.href} not in stylesheetAstObjects!`)
+            // }
+            // if (!Object.keys(stylesheetAstObjects[link.href]).length) {
+            //   // If the 'stylesheetAstObjects[link.href]' thing is an
+            //   // empty object, simply skip this link.
+            //   return
+            // }
+            hrefs.push(link.href)
+          }
+        })
+        return {
+          html: document.documentElement.outerHTML,
+          hrefs
+        }
+      })
 
-    const htmlNetworkIdle = evalNetworkIdle.html
-    doms.push(cheerio.load(htmlNetworkIdle))
-    evalNetworkIdle.hrefs.forEach(href => {
-      allHrefs.add(href)
-    })
+      const htmlNetworkIdle = evalNetworkIdle.html
+      doms.push(cheerio.load(htmlNetworkIdle))
+      evalNetworkIdle.hrefs.forEach(href => {
+        allHrefs.add(href)
+      })
 
-    // We can close the browser now that all URLs have been opened.
-    if (!options.browser) {
-      browser.close()
+      // We can close the browser now that all URLs have been opened.
+      if (!options.browser) {
+        browser.close()
+        browser = null
```

Comment:
This needs a comment that explains **why** the `browser` variable is changed.

Reply:
I suppose there is better way to do this, what is basically required is `finally` to make sure we close browser, instead I'm using `browser=null` as a flag that browser already closed. Probably need to fix this